agroal-api/src/main/java/io/agroal/api/cache/LocalConnectionCache.java

```diff
@@ -0,0 +1,137 @@
+                    threadLocal.set( acquirable );
+                }
+            }
+
+            @Override
+            public void reset() {
+                threadLocal = new ThreadLocal<>();
+            }
+        };
+    }
+
+    /**
+     * A local cache that stores up to a number of connections.
+     */
+    static ConnectionCache fixed(int size) {
+        return new ConnectionCache() {
+
+            @SuppressWarnings( "ThreadLocalNotStaticFinal" )
+            private ThreadLocal<Acquirable[]> threadLocal;
+
+            { // instance initializer
+                reset();
+            }
+
+            @Override
+            public Acquirable get() {
+                Acquirable[] cacheArray = threadLocal.get();
+                for ( int i = cacheArray.length; i > 0; ) { // iterate from last to first
+                    Acquirable element = cacheArray[--i];
+                    if ( element != null ) {
+                        cacheArray[i] = null;
+                        if ( element.acquire() ) {
+                            return element;
+                        }
+                    }
+                }
+                return null;
+            }
+
+            @Override
+            public void put(Acquirable acquirable) {
+                Acquirable[] cacheArray = threadLocal.get();
+                int last = cacheArray.length - 1;
+                if ( cacheArray[last] != null ) { // replace last
+                    cacheArray[last] = acquirable;
+                } else {
+                    int i = 0;
+                    while ( cacheArray[i++] != null ) ; // find first free slot
+                    cacheArray[--i] = acquirable;
+                }
+            }
+
+            @Override
+            public void reset() {
+                threadLocal = ThreadLocal.withInitial( () -> new Acquirable[size] );
+            }
+        };
+    }
+
+    /**
+     * A local cache that stores all connections
+     */
+    static ConnectionCache full() {
+        return new ConnectionCache() {
+
+            @SuppressWarnings( "ThreadLocalNotStaticFinal" )
+            private ThreadLocal<Deque<Acquirable>> threadLocal;
+
+            { // instance initializer
+                reset();
+            }
+
+            @Override
+            public Acquirable get() {
+                Deque<Acquirable> queue = threadLocal.get();
+
+                for ( int i = queue.size(); i > 0; i-- ) {
+                    Acquirable acquirable = queue.removeFirst();
+                    if ( acquirable.acquire() ) {
+                        return acquirable;
+                    } else if ( acquirable.isAcquirable() ) {
+                        queue.addLast( acquirable );
+                    }
+                }
+                return null;
+            }
+
+            @Override
+            public void put(Acquirable acquirable) {
+                threadLocal.get().addFirst( acquirable );
```

Comment:
Shouldn't this have an isAcquirable check?

Reply:
yes, this needs to be fixed.